contracts/bridge/ABridge.sol

```diff
@@ -2,236 +2,160 @@
 pragma solidity ^0.8.21;
 
 import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
 import {MessageHashUtils} from "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
 import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
 import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {EnumerableMap} from "@openzeppelin/contracts/utils/structs/EnumerableMap.sol";
 
 import {IBridge} from "../interfaces/bridge/IBridge.sol";
+import {IHandler} from "../interfaces/bridge/IHandler.sol";
 
-import {AERC20Handler} from "./handlers/AERC20Handler.sol";
-import {AERC721Handler} from "./handlers/AERC721Handler.sol";
-import {AERC1155Handler} from "./handlers/AERC1155Handler.sol";
-import {ANativeHandler} from "./handlers/ANativeHandler.sol";
+import {IBatcher, Batcher} from "./batcher/Batcher.sol";
 
 /**
  * @notice The Bridge module
  *
- * The Bridge contract facilitates the permissioned transfer of assets (ERC-20, ERC-721, ERC-1155, Native)
+ * The Bridge contract facilitates the permissioned transfer of assets and/or arbitrary messages
  * between two (or more) EVM blockchains.
  *
- * To utilize the Bridge effectively, instances of this contract must be deployed on both base and destination chains,
- * accompanied by the setup of trusted back ends to act as signers.
+ * To utilize the Bridge effectively, instances of this contract must be deployed on both base
+ * and destination chains, accompanied by the setup of trusted back ends to act as signers.
+ * The back end signatures are checked only upon redemption.
  *
- * The Bridge contract supports both "liquidity pool" and "mint-and-burn" methods for managing assets.
- * The back end signatures are checked only upon token withdrawals. If "mint-and-burn" method is used,
- * the ERC-20 tokens are required to support ERC-7802 interface.
+ * Each asset type is mapped to a handler (ERC-20, Native, Message) contract.
+ * Handlers implement the specific dispatch/redeem logic.
  *
- * The Bridge is also suitable for bridged USDC tokens, utilizing their interface
- * (https://github.com/circlefin/stablecoin-evm/blob/master/doc/bridged_USDC_standard.md).
+ * During redeem, handlers may forward execution to the `IBatcher` contract for multi-step operations.
```

Comment:
Let's add a few lines saying that "the bridge never executes batches directly and does this via a designated "batcher" contract that can be overridden. Users are advised to construct tx batches in a way that they are not front-runnable. E.g. USDT approve should be reset first."